File: src/macro/kvstore_node/workload/workload.js
const utils = require('../utils');
const generator = require('../generators/generators');
const defaults = require('./defaults');
const {ConstantGenerator, UniformGenerator, ZipfianGenerator, CounterGenerator, ScrambledZipfianGenerator,
    SkewedZipfianGenerator, DiscreteGenerator
} = require("../generators/generators");



class CoreWorkload {
    field_count_ = 0;
    read_all_fields_ = false;
    write_all_fields_ = false;
    field_len_generator_ = undefined;
    key_generator_ = undefined;
    key_chooser_ = undefined;
    field_chooser_ = undefined;
    scan_len_chooser_ = undefined;
    insert_key_sequence_;
    ordered_inserts_ = true;
    record_count_ = 0;

    op_chooser_ = new DiscreteGenerator();
    constructor (){
        this.insert_key_sequence_ = new CounterGenerator(3);
    }

    Init(properties) {
        this.table_name_ = properties.GetProperty('table', 'usertable');
        this.field_count_ = properties.GetProperty('fieldcount', 10);
        this.field_len_generator_ = this.GetFieldLenGenerator(properties);

        const read_proportion = properties.GetProperty('readproportion', 0.95);
        const update_proportion = properties.GetProperty('updateproportion', 0.05);
        const insert_proportion = properties.GetProperty('insertproportion', 0.0);
        const scan_proportion = properties.GetProperty('scanproportion', 0.0);
        const readmodifywrite_proportion = properties.GetProperty('readmodifywriteproportion', 0.0);

        this.record_count_ = properties.GetProperty('recordcount', 'operationcount');
        const request_dist = properties.GetProperty('requestdistribution', 'uniform');
        const max_scan_len = properties.GetProperty('maxscanlength', 1000);
        const scan_len_dist = properties.GetProperty('scanlengthdistribution', 'uniform');
        const insert_start = properties.GetProperty('insertstart', 0);

        this.read_all_fields_ = properties.GetProperty('readallfields', true);
        this.write_all_fields_ = properties.GetProperty('writeallfields', false);

        this.ordered_inserts_ = properties.GetProperty('insertorder', 'hashed') !== 'hashed';

        this.key_generator_ = new CounterGenerator(insert_start);

        if (read_proportion > 0) this.op_chooser_.AddValue('READ', read_proportion);
        if (update_proportion > 0) this.op_chooser_.AddValue('UPDATE', update_proportion);
        if (insert_proportion > 0) this.op_chooser_.AddValue('INSERT', insert_proportion);
        if (scan_proportion > 0) this.op_chooser_.AddValue('SCAN', scan_proportion);
        if (readmodifywrite_proportion > 0) this.op_chooser_.AddValue('READMODIFYWRITE', readmodifywrite_proportion);

        this.insert_key_sequence_.Set(this.record_count_);

        switch (request_dist) {
            case 'uniform': {
                this.key_chooser_ = new UniformGenerator(0, this.record_count_ - 1);
                break;
            }
            case 'zipfian': {
                const op_count = properties.GetProperty('operationcount');
                const new_keys = Math.floor(op_count * insert_proportion * 2);
                this.key_chooser_ = new ScrambledZipfianGenerator(this.record_count_ + new_keys);
                break;
            }
            case 'latest': {
                this.key_chooser_ = new SkewedZipfianGenerator(this.insert_key_sequence_);
                break;
            }
            default: {
                throw `Unknown distribution ${request_dist}`
            }
        }

        this.field_chooser_ = new UniformGenerator(0, this.field_count_ - 1);

        switch (scan_len_dist) {
            case 'uniform': {
                this.scan_len_chooser_ = new UniformGenerator(1, max_scan_len);
                break;
            }
            case 'zipfian': {
                this.scan_len_chooser_ = new ZipfianGenerator(1, max_scan_len);
                break;
            }
            default: {
                throw `Not allowed distribution: ${scan_len_dist}`;
            }
        }
    }

    NextTable() {
        return this.table_name_;
    }

    NextOperation() {
        return this.op_chooser_.Next();
    }

    NextScanLength() {
        return this.scan_len_chooser_.Next();
    }

    read_all_fields() {
        return this.read_all_fields_;
    }

    write_all_fields() {
        return this.write_all_fields_;
    }

    NextSequenceKey() {
        const key_num = this.key_generator_.Next();
        return this.BuildKeyName(key_num);
    }

    NextTransactionKey() {
        let key_num;
        do {
            key_num = this.key_chooser_.Next();
        } while(key_num > this.insert_key_sequence_.Last());
        return this.BuildKeyName(key_num);
    }

    BuildKeyName(key_num) {
        if (!this.ordered_inserts_) {
            key_num = utils.Hash(key_num);
        }
        return `user${key_num}`;
    }

    NextFieldName() {
        return `field${this.field_chooser_.Next()}`;
    }

    BuildValues() {
        let values = [];
        for (let i=0; i < this.field_count_; i += 1) {
            values.push([`field${i}`,
                `${this.field_len_generator_.Next()}${utils.RandomPrintChar()}`]);
        }
        return values;
    }

    BuildUpdate(update) {
        update.push(
            [this.field_len_generator_.Next(),
            utils.RandomPrintChar,]
        );
        return update;
    }

    GetFieldLenGenerator(properties) {
        const field_len_dist = properties.GetProperty('field_len_dist', 'constant');
        const field_len = properties.GetProperty('fieldlength', 100);

        switch (field_len_dist) {
            case 'constant': {
                return new ConstantGenerator(field_len);
            }
            case 'uniform': {
                return new UniformGenerator(1, field_len);
            }
            case 'zipfian': {
                return new ZipfianGenerator(1, field_len);
            }
            default : {
                throw `Unknown generator: ${field_len_dist}`
            }
        }
    }
}

module.exports = {
    Workload: CoreWorkload,
}

File: src/macro/rust-driver/props/src/workload.rs
use std::cell::RefCell;
use std::rc::Rc;
use std::str::FromStr;
use generator::constant::ConstGenerator;
use generator::counter::CounterGenerator;
use generator::discrete::{DiscreteGenerator, Operation, ValueDoublePair};
use generator::generator::Generator;
use generator::scrambled_zipfian::ScrambledZipfianGenerator;
use generator::skewed_latest::SkewedLatestGenerator;
use generator::uniform::UniformGenerator;
use generator::zipfian::ZipfianGenerator;
use crate::properties::Properties;
use generator::{random_print_char, fnv_hash_64};
use utils::Pair;

#[derive(Debug)]
pub enum GeneratorType {
    Constant(ConstGenerator),
    Counter(CounterGenerator),
    Uniform(UniformGenerator),
    Scrambled(ScrambledZipfianGenerator),
    Skewed(SkewedLatestGenerator),
    Zipfian(ZipfianGenerator),
    None
}

#[derive(Debug)]
pub struct Workload
{
    table_name_: Option<String>,
    field_count_: u64,
    read_all_fields_: bool,
    write_all_fields_: bool,
    field_len_generator_: GeneratorType,
    key_generator_: GeneratorType,
    op_chooser_: DiscreteGenerator<Operation>,
    key_chooser_: GeneratorType,
    field_chooser_: GeneratorType,
    scan_len_chooser_: GeneratorType,
    insert_key_sequence_: Rc<RefCell<CounterGenerator>>,
    ordered_inserts_: bool,
    record_count_: u64,
}

impl Workload {
    pub fn new() -> Self {
        Self {
            table_name_: None,
            field_count_: 0,
            read_all_fields_: false,
            write_all_fields_: false,
            field_len_generator_: GeneratorType::None,
            key_generator_: GeneratorType::None,
            op_chooser_: DiscreteGenerator::new(),
            key_chooser_: GeneratorType::None,
            field_chooser_: GeneratorType::None,
            scan_len_chooser_: GeneratorType::None,
            insert_key_sequence_: Rc::new(RefCell::new(CounterGenerator::new(3))),
            ordered_inserts_: true,
            record_count_: 0,
        }
    }

    pub fn init(&mut self, props: &Properties) {
        self.table_name_ = Some(props.get_or_default("table").to_string());
        self.field_count_ = u64::from_str(props.get_or_default("fieldcount")).unwrap();
        self.field_len_generator_ = Self::get_field_generator(props);

        let read_proportion = f64::from_str(props.get_or_default("readproportion")).unwrap();
        let update_proportion = f64::from_str(props.get_or_default("updateproportion")).unwrap();
        let insert_proportion = f64::from_str(props.get_or_default("insertproportion")).unwrap();
        let scan_proportion = f64::from_str(props.get_or_default("scanproportion")).unwrap();
        let read_modify_write_proportion = f64::from_str(props.get_or_default("readmodifywriteproportion")).unwrap();
        self.record_count_ = u64::from_str(&props["recordcount"]).unwrap();

        let request_dist = props.get_or_default("requestdistribution");
        let max_scan_len = u64::from_str(props.get_or_default("maxscanlength")).unwrap();
        let scan_len_dist = props.get_or_default("scanlengthdistribution");
        let insert_start = u64::from_str(props.get_or_default("insertstart")).unwrap();

        self.read_all_fields_ = bool::from_str(props.get_or_default("readallfields")).unwrap();
        self.write_all_fields_ = bool::from_str(props.get_or_default("writeallfields")).unwrap();

        self.ordered_inserts_ = props.get_or_default("insertorder") != "hashed";

        self.key_generator_ = GeneratorType::Counter(CounterGenerator::new(insert_start));

        if read_proportion > 0f64 {
            self.op_chooser_.add_value(Operation::Read, read_proportion);
        }

        if update_proportion > 0f64 {
            self.op_chooser_.add_value(Operation::Update, update_proportion);
        }

        if insert_proportion > 0f64 {
            self.op_chooser_.add_value(Operation::Insert, insert_proportion);
        }

        if scan_proportion > 0f64 {
            self.op_chooser_.add_value(Operation::Scan, scan_proportion);
        }

        if read_modify_write_proportion > 0f64 {
            self.op_chooser_.add_value(Operation::ReadModifyWrite, read_modify_write_proportion);
        }

        self.insert_key_sequence_.borrow_mut().set(self.record_count_);

        self.key_chooser_ = match request_dist {
            "uniform" => GeneratorType::Uniform(UniformGenerator::new(0, self.record_count_ - 1)),
            "zipfian" => {
                let op_count = u64::from_str(&props["operationcount"]).unwrap();
                let new_keys = (op_count as f64 * insert_proportion * 2f64) as u64;
                GeneratorType::Scrambled(ScrambledZipfianGenerator::new_single(self.record_count_ + new_keys))
            }
            "latest" => GeneratorType::Skewed(SkewedLatestGenerator::new(Rc::clone(&self.insert_key_sequence_))),
            _ => GeneratorType::None
        };

        self.field_chooser_ = GeneratorType::Uniform(UniformGenerator::new(0, self.field_count_ - 1));

        self.scan_len_chooser_ = match scan_len_dist {
            "uniform" => GeneratorType::Uniform(UniformGenerator::new(1, max_scan_len)),
            "zipfian" => GeneratorType::Zipfian(ZipfianGenerator::new_multiple(1, max_scan_len, ZipfianGenerator::K_ZIPFIAN_CONST)),
            _ => GeneratorType::None
        };
    }

    fn get_field_generator(props: &Properties) -> GeneratorType {
        let field_len_dist = props.get_or_default("field_len_dist");
        let field_len = u64::from_str(props.get_or_default("fieldlength")).unwrap();

        match field_len_dist {
            "constant" => {
                GeneratorType::Constant(ConstGenerator::new(field_len))
            }
            "uniform" => {
                GeneratorType::Uniform(UniformGenerator::new(1, field_len))
            }
            "zipfian" => {
                GeneratorType::Zipfian(ZipfianGenerator::new_multiple(1, field_len, ZipfianGenerator::K_ZIPFIAN_CONST))
            }
            _ => {
                GeneratorType::None
            }
        }
    }

    pub fn read_all_fields(&self) -> bool {
        self.read_all_fields_
    }

    pub fn write_all_fields(&self) -> bool {
        self.write_all_fields_
    }

    pub fn next_scan_length(&mut self) -> u64 {
        match &mut self.scan_len_chooser_ {
            GeneratorType::Uniform(uniform) => uniform.next(),
            GeneratorType::Zipfian(zipfian) => zipfian.next(),
            _ => 0
        }
    }

    pub fn next_table(&self) -> String {
        let mut result;
        if let Some(table_name) = &self.table_name_ {
            result = table_name.clone();
        }else {
            result = "".to_string();
        }
        result
    }

    pub fn build_values(&mut self, values: &mut Vec<Pair>) {
        for idx in 0..self.field_count_ {
            let value = match &mut self.field_len_generator_ {
                GeneratorType::Constant(constant) => constant.next(),
                GeneratorType::Uniform(uniform) => uniform.next(),
                GeneratorType::Zipfian(zipfian) => zipfian.next(),
                _ => 0
            };
            values.push(Pair {
                key: format!("field{}", idx).to_string(),
                value: format!("{}{}", value, random_print_char()),
            })
        }
    }

    pub fn next_sequence_key(&mut self) -> String {
        let mut key_num = 0;
        if let GeneratorType::Counter(counter) = &mut self.key_generator_ {
            key_num = counter.next();
        }
        self.build_key_name(key_num)
    }

    pub fn next_transaction_key(&mut self) -> String {
        let mut key_num;
        loop {
            key_num = match &mut self.field_len_generator_ {
                GeneratorType::Uniform(uniform) => uniform.next(),
                GeneratorType::Zipfian(zipfian) => zipfian.next(),
                _ => 0
            };
            if key_num > self.insert_key_sequence_.borrow_mut().last() {
                break;
            }
        }
        self.build_key_name(key_num)
    }

    pub fn build_key_name(&mut self, mut key_num: u64) -> String {
        if !self.ordered_inserts_ {
            key_num = fnv_hash_64(key_num);
        }
        format!("user{}", key_num)
    }

    pub fn next_field_name(&mut self) -> String {
        let mut field_name: String = "".to_string();
        if let GeneratorType::Uniform(uniform) = &mut self.field_chooser_ {
            field_name = format!("field{}",uniform.next());
        }
        field_name
    }

    pub fn next_operation(&mut self) -> Operation {
        self.op_chooser_.next()
    }

    pub fn build_update(&mut self, update: &mut Vec<Pair>) {

        let value = match &mut self.field_len_generator_ {
            GeneratorType::Constant(constant) => constant.next(),
            GeneratorType::Uniform(uniform) => uniform.next(),
            GeneratorType::Zipfian(zipfian) => zipfian.next(),
            _ => 0
        };
        update.push(Pair {
            key: self.next_field_name(),
            value: format!("{}{}", value, random_print_char()),
        });
    }
}